doc: Add initial documentation for N-API

mhdawson opened the pull request:
Add the initial documentation for the N-API

[**Rendered**](https://github.com/mhdawson/io.js/blob/napi-doc/doc/api/n-api.md)

This PR is a result of work in the abi-stable-node repo:
https://github.com/nodejs/abi-stable-node/tree/doc,
with this PR being the cumulative work on the documentation
sections in that repo with the following contributors
in alphabetical order:

    Author: Arunesh Chandra <arunesh.chandra@microsoft.com>
    Author: Gabriel Schulhof <gabriel.schulhof@intel.com>
    Author: Hitesh Kanwathirtha <hiteshk@microsoft.com>
    Author: Jason Ginchereau <jasongin@microsoft.com>
    Author: Michael Dawson <michael_dawson@ca.ibm.com>
    Author: Sampson Gao <sampsong@ca.ibm.com>
    Author: Taylor Woll <taylor.woll@microsoft.com>

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [X] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [X] documentation is changed or added
- [X] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-guidelines)

##### Affected core subsystem(s)
doc, n-api

jasongin commented on doc/api/_toc.md:
```diff
@@ -24,6 +24,7 @@
 * [HTTP](http.html)
 * [HTTPS](https.html)
 * [Modules](modules.html)
+* [N-API](n-api.html)
```
Should be titled something like "C/C++ Addons (N-API)", so that it is adjacent to the other "C/C++ Addons" topic?

mhdawson replied:
makes sense to me, will update once we have more comments.

vsemozhetbyt commented on doc/api/addons.md:
```diff
@@ -213,6 +213,24 @@ Addon developers are recommended to use to keep compatibility between past and
+or [Native Abstractions for Node.js][] APIs, the functions available
+in the N-API are used instead.
+
+The functions available and how to use them are document in the 
```
documented?

mhdawson replied:
fixed

cjihrig commented on doc/api/addons.md:
```diff
@@ -213,6 +213,24 @@ Addon developers are recommended to use to keep compatibility between past and
+
+## N-API
+
+N-API is a newer API for building native Addons.  It is independent from
```
I would drop `newer` here. Otherwise, it will probably claim to be new for years to come.

cjihrig replied:
I don't know if we standardized anywhere, but you have two spaces after some sentences, and one period after others.

mhdawson replied:
Fixed.

gibfahn replied:
We should standardise, probably on 1.

cjihrig commented on doc/api/addons.md:
```diff
@@ -213,6 +213,24 @@ Addon developers are recommended to use to keep compatibility between past and
+changes in the underlying JavaScript engine and allow modules
+compiled for one version to run on later versions of Node.js without
+recompilation.  Addons are built/packaged with the same approach/tools
+outlined in this document(node-gyp, etc.). The only difference is the
```
Add a space after "document"

mhdawson replied:
done

cjihrig commented on doc/api/addons.md:
```diff
@@ -213,6 +213,24 @@ Addon developers are recommended to use to keep compatibility between past and
+outlined in this document(node-gyp, etc.). The only difference is the
+set of APIs that are used by the native code.  Instead of using the V8
+or [Native Abstractions for Node.js][] APIs, the functions available
+in the N-API are used instead.
```
This sentence begins and ends with "instead." I think the second one can be dropped.

mhdawson replied:
done

cjihrig commented on doc/api/n-api.md:
```diff
@@ -0,0 +1,3151 @@
+outlined the second titled  [C/C++ Addons](addons.html).
+The only difference is the set of APIs that are used by the native code.
+Instead of using the V8 or [Native Abstractions for Node.js][] APIs,
+the functions available in the N-API are used instead.
```
Same comments in these first two paragraphs apply.

mhdawson replied:
done

cjihrig commented on doc/api/n-api.md:
```diff
@@ -0,0 +1,3151 @@
+the functions available in the N-API are used instead.
+
+
+APIs exposed by the N-API surface are generally used to create and manipulate
```
"exposed by the N-API surface" -> "exposed by N-API"

mhdawson replied:
done

cjihrig commented on doc/api/n-api.md:
```diff
@@ -0,0 +1,3151 @@
+in the ECMA262 Language Specification. The APIs have the following 
+properties:
+- All N-API calls return a status code of type `napi_status`. This
+  status can be used to determine whether the API call succeeded or failed.
```
"can be used to determine" -> "indicates"

mhdawson replied:
done

cjihrig commented on doc/api/n-api.md:
```diff
@@ -0,0 +1,3151 @@
+properties:
+- All N-API calls return a status code of type `napi_status`. This
+  status can be used to determine whether the API call succeeded or failed.
+- The actual value being returned by the API is returned in an out parameter.
```
Maybe rewrite as "The API's return value is passed via an out parameter."

mhdawson replied:
done

cjihrig commented on doc/api/n-api.md:
```diff
@@ -0,0 +1,3151 @@
+  `napi_value`.
+- In case of an error status code, additional information can be obtained 
+  using `napi_get_last_error_info`. More information can be found in the error
+  handling section [Error Handling][Error Handling].
```
Should this be `[Error Handling][]`?

mhdawson replied:
Fixed

cjihrig commented on doc/api/n-api.md:
```diff
@@ -0,0 +1,3151 @@
+* [Object Wrap][]
+* [Aynchronous Operations][]
+
+The N-API is a C based API in order to allow us to ensure ABI stability
```
"C based" -> "C"
"in order to allow us to ensure" -> "that ensures"

mhdawson replied:
done

cjihrig commented on doc/api/n-api.md:
```diff
@@ -0,0 +1,3151 @@
+across Node.js versions and different compiler levels.  However, we also
+understand that a C++ API can be easier/faster to use in many
+cases.  To support these cases we expect there to be one or more
+C++ wrapper modules that provide an in-lineable C++ API that results in
```
I don't think in-lineable has a hyphen.

cjihrig replied:
This whole sentence is kind of confusing. Consider breaking into multiple sentences?

mhdawson replied:
fxied

mhdawson replied:
done

cjihrig commented on doc/api/n-api.md:
````diff
@@ -0,0 +1,3151 @@
+```
+
+#### Members
+- error_message: UTF8-encoded string containing a VM-neutral description of 
````
Should the members be in backticks?

mhdawson replied:
We did not do that for any other headers.

cjihrig commented on doc/api/n-api.md:
```diff
@@ -0,0 +1,3151 @@
+
+#### Members
+- error_message: UTF8-encoded string containing a VM-neutral description of 
+  the error
```
Some of these items have periods at the end, others don't.

mhdawson replied:
fixed.

cjihrig commented on doc/api/n-api.md:
```diff
@@ -0,0 +1,3151 @@
+  not implemented for any VM.
+- error_code: the N-API status code that originated with the last error
+
+#### Additional Information
```
Maybe drop this header and say:

`See the [Error Handling][] section for additional information.`

mhdawson replied:
done

cjihrig commented on doc/api/n-api.md:
```diff
@@ -0,0 +1,3151 @@
+This is an abstraction used to control and modify the lifetime of objects 
+created within a particular scope. In general, N-API values are created within
+the context of a handle scope. When a native method is called from 
+JavaScript, a default handle scope will exist- if the user does not explicitly
```
Instead of `exist-`, can you just end the sentence and start a new one.

mhdawson replied:
done

cjihrig commented on doc/api/n-api.md:
```diff
@@ -0,0 +1,3151 @@
+
+#### *napi_callback*
+Function pointer type for user-provided native functions which are to be 
+exposed to JavaScript via N-API. Your callback function should satisfy the 
```
Can you drop the pronoun and say something like "Callback functions must have the following signature:"

mhdawson replied:
Done

cjihrig commented on doc/api/n-api.md:
````diff
@@ -0,0 +1,3151 @@
+
+```C
+typedef void (*napi_async_execute_callback)(napi_env env,
+                                            void* data);
````
I think this fits on the previous line.

mhdawson replied:
fixed

cjihrig commented on doc/api/n-api.md:
```diff
@@ -0,0 +1,3151 @@
+
+#### napi_async_complete_callback
+
+Function pointer used with the functions that support asynchronous
```
"the functions" -> "functions.

mhdawson replied:
done

cjihrig commented on doc/api/n-api.md:
```diff
@@ -0,0 +1,3151 @@
+#### napi_async_complete_callback
+
+Function pointer used with the functions that support asynchronous
+operations. Your callback function should statisfy the following signature:
```
Please drop the pronoun. I'll stop mentioning it :-)

mhdawson replied:
Done

cjihrig commented on doc/api/n-api.md:
```diff
@@ -0,0 +1,3151 @@
+
+## Error Handling
+N-API uses both return values and Javascript exceptions for error handling.
+The sections which follow explain the approach for each case.
```
"sections which follow" -> "following sections"

mhdawson replied:
done

cjihrig commented on doc/api/n-api.md:
```diff
@@ -0,0 +1,3151 @@
+The full set of possible napi_status values is defined
+in `napi_api_types.h`.
+
+Thr `napi_status` return value provides a vm-independent representation of
```
Thr -> The

cjihrig replied:
Should "vm" be "VM" throughout?

mhdawson replied:
Fixed

mhdawson replied:
changed

cjihrig commented on doc/api/n-api.md:
```diff
@@ -0,0 +1,3151 @@
+more detailed information, including a string representing the error as well as
+vm(engine)-specific information.  **NOTE:**  you should not rely on the
+content or format of any of the extended information as it is
+not subject to SemVer and may change at any time. It is intended
```
Is this something we have control over? If not, it might be useful to say so.

mhdawson replied:
We do have some control in that the information will be put together by the 'port' of the N-API implementation for each VM (ex V8, Chakra) which is under out control.  However the information that makes sense may vary across VM's and it may incorporate information under the control of the VM.

cjihrig commented on doc/api/n-api.md:
````diff
@@ -0,0 +1,3151 @@
+```C
+NODE_EXTERN const napi_extended_error_info* napi_get_last_error_info();
+```
+and the format of the `napi_extended_error_info` structure is as follows:
````
Maybe "and the" -> "The"

mhdawson replied:
done

cjihrig commented on doc/api/n-api.md:
````diff
@@ -0,0 +1,3151 @@
+```
+
+`napi_get_last_error_info()` returns the information for the last
+N-API call that was made and the  `error_code` will match the
````
Maybe start a new sentence at "and the"

mhdawson replied:
done

cjihrig commented on doc/api/n-api.md:
````diff
@@ -0,0 +1,3151 @@
+napi_is_exception_pending()
+```
+
+and then check the value of `result`.
````
This could probably be combined with the last sentence of the previous paragraph.

mhdawson replied:
done

cjihrig commented on doc/api/n-api.md:
```diff
@@ -0,0 +1,3151 @@
+and then return to JavaScript where the exception can be handled.
+
+The second approach is to try to handle the exception. There will be cases
+where the native code can 'catch' the exception, take the appropriate action,
```
Does catch need to be in single quotes?

mhdawson replied:
does flow better without, removed.

cjihrig commented on doc/api/n-api.md:
````diff
@@ -0,0 +1,3151 @@
+napi_get_and_clear_last_exception()
+```
+
+On success, result will contain the handle to the JavaScript `Error` object.
````
Is this guaranteed to be an `Error` object, or can it be whatever was thrown by JS?

mhdawson replied:
That is a good point, it will be whatever was thrown.

cjihrig commented on doc/api/n-api.md:
````diff
@@ -0,0 +1,3151 @@
+```
+
+The following utility functions are also available in case your native
+code needs to create an error object:
````
error -> `Error`

mhdawson replied:
fixed

vsemozhetbyt commented on doc/api/n-api.md:
```diff
@@ -0,0 +1,3151 @@
+recompilation.
+
+Addons are built/packaged with the same approach/tools
+outlined the second titled  [C/C++ Addons](addons.html).
```
the second -> in the section titled?

mhdawson replied:
Fixed.

vsemozhetbyt commented on doc/api/n-api.md:
```diff
@@ -0,0 +1,3151 @@
+                                              napi_value value,
+                                              int initial_refcount,
+                                              napi_ref* result);
+##### Parameters
```
missing backticks above

mhdawson replied:
Fixed

vsemozhetbyt commented on doc/api/n-api.md:
````diff
@@ -0,0 +1,3151 @@
+```C
+napi_open_escapable_handle_scope()
+napi_close_escapable_handle_scope()
+
````
missing backticks above

mhdawson replied:
Fixed

vsemozhetbyt commented on doc/api/n-api.md:
```diff
@@ -0,0 +1,3151 @@
+##### Description
+This represents the underlying binary scalar datatype of the TypedArray.
+Elements of this enum correspond to 
+[Section 22.2] (https://tc39.github.io/ecma262/#sec-typedarray-objects)
```
buggy space

mhdawson replied:
Fixed

vsemozhetbyt commented on doc/api/n-api.md:
```diff
@@ -0,0 +1,3151 @@
+##### Parameters
+- `[in] env `: The environment that the API is invoked under
+- `[in] msg`: `napi_value` to be checked
+- `[out] `result`: boolean value. Set to true if `napi_value` represents
```
wrong backtick

mhdawson replied:
fixed

vsemozhetbyt commented on doc/api/n-api.md:
```diff
@@ -0,0 +1,3151 @@
+
+#### Description
+This API implements the abstract operation ToString as defined in
+[Section 7.1.13](http
```
data' -> data`

vsemozhetbyt replied:
data' -> data`

mhdawson replied:
fixed

vsemozhetbyt commented:
Till #12554 fixed, maybe it is reasonable to split very big docs into several parts for review time.

mhdawson commented:
Did a pass 

* replacing spaces after period with one space
* moving use of your

mhdawson commented:
@cjihrig @vsemozhetbyt pushed commits to address all comments so far.

mhdawson commented:
I'm happy to split up into separate PRs or review in this single PR.  I'll leave it up to the reviewers to comment on which they think in best.

Fishrock123 commented on doc/api/n-api.md:
```diff
@@ -0,0 +1,3150 @@
+the functions available in the N-API are used.
+
+
+APIs exposed by N-AP are generally used to create and manipulate
```
N-API

mhdawson replied:
fixed

Fishrock123 commented on doc/api/n-api.md:
```diff
@@ -0,0 +1,3150 @@
+
+
+N-API is an API for building native Addons. It is independent from
+the underlying JavaScript runtime (ex v8) and is maintained as part of
```
`V8`

mhdawson replied:
done

Fishrock123 commented on doc/api/addons.md:
```diff
@@ -213,6 +213,24 @@ Addon developers are recommended to use to keep compatibility between past and
+## N-API
+
+N-API is an API for building native Addons. It is independent from
+the underlying JavaScript runtime (ex v8) and is maintained as part of
```
`V8`

mhdawson replied:
done

Fishrock123 commented:
Added a `rendered` link in the OP, points to the branch, not a specific commit: https://github.com/mhdawson/io.js/blob/napi-doc/doc/api/n-api.md

Fishrock123 commented:
Can we get rid of the `Signature`, `Parameters`, `Return value`, `Description` mini-headers? They take up a lot of space in the long document and are unnecessary. We could move it to be [more like a JS api example](https://nodejs.org/dist/latest-v7.x/docs/api/fs.html#fs_fs_createwritestream_path_options), for consistency and readability.

Consider this comparison:

----
#### napi_close_handle_scope
##### Signature

```C
NODE_EXTERN napi_status napi_close_handle_scope(napi_env e,
                                                napi_handle_scope scope);
```
##### Parameters

- `[in] env `: The environment that the API is invoked under
- `[in] scope`: `napi_value` representing the scope to be closed

##### Return value

- `napi_ok` if the API succeeded.

##### Description

This API closes the scope passed in. Scopes must be closed in the
reverse order from which they were created.

----
vs

----

#### napi_close_handle_scope
```C
NODE_EXTERN napi_status napi_close_handle_scope(napi_env e,
                                                napi_handle_scope scope);
```
- `[in] env `: The environment that the API is invoked under
- `[in] scope`: `napi_value` representing the scope to be closed

Returns `napi_ok` if the API succeeded.

This API closes the scope passed in. Scopes must be closed in the
reverse order from which they were created.

----

Fishrock123 commented:
Also, can we add `added-in` placeholders to the API methods?

Fishrock123 commented on doc/api/n-api.md:
```diff
@@ -0,0 +1,3150 @@
+
+##### Description
+This API closes the scope passed in. Scopes must be closed in the
+reverse order from which they were created.o
```
`created.o` is a typo

mhdawson replied:
fixed

addaleax commented on doc/api/n-api.md:
```diff
@@ -0,0 +1,3150 @@
+
+The N-API is a C API that ensures ABI stability
+across Node.js versions and different compiler levels. However, we also
+understand that a C++ API can be easier/faster to use in many
```
This paragraph seems to be talking about C++ wrappers around N-API, so I would strike `faster` here – they would be easier but they would inherently be limited to being as fast as N-API at best

mhdawson replied:
meant faster in terms of development not speed, but remove to avoid confusion

addaleax commented on doc/api/n-api.md:
```diff
@@ -0,0 +1,3150 @@
+users to manage the lifetimes of JavaScript values, including defining their 
+minimum lifetimes explicitly.
+
+For more details, review the [Object Lifetime Management section](TODO://path).
```
TODO?

mhdawson replied:
fixed

addaleax commented on doc/api/n-api.md:
```diff
@@ -0,0 +1,3150 @@
+The `napi_status` return value provides a VM-independent representation of
+the error which occurred. In some cases it is useful to be able to get
+more detailed information, including a string representing the error as well as
+VM (engine)-specific information. **NOTE:**  you should not rely on the
```
Notes should be on separate paragraphs and not be all uppercase, followed by a capitalized sentence, i.e. `*Note:* You should …`

mhdawson replied:
Done, and fixed other instances as well.

addaleax commented on doc/api/n-api.md:
````diff
@@ -0,0 +1,3150 @@
+In order to retrieve this information, the following method is provided:
+
+```C
+NODE_EXTERN const napi_extended_error_info* napi_get_last_error_info();
````
This is not the actual signature, though?

mhdawson replied:
No, to keep thing  shorter and because we have the full signature/details in the later sections moved it from these earlier references.

mhdawson replied:
Maybe adding ... for the parameters would help ?

addaleax commented on doc/api/n-api.md:
````diff
@@ -0,0 +1,3150 @@
+```
+
+`napi_get_last_error_info()` returns the information for the last
+N-API call that was made. Tthe  `error_code` will match the
````
typo: `Tthe` and double space after it

mhdawson replied:
fixed

addaleax commented on doc/api/n-api.md:
```diff
@@ -0,0 +1,3150 @@
+
+##### Description
+This API retrieves a `napi_extended_error_info` structure with information
+about the last error tha occured.
```
typo: tha

mhdawson replied:
fixed

addaleax commented on doc/api/n-api.md:
```diff
@@ -0,0 +1,3150 @@
+
+### Exceptions
+
+Any N-API function may result in a pending JavaScript exception. This is
```
s/function/function call/?

mhdawson replied:
changed

addaleax commented on doc/api/n-api.md:
```diff
@@ -0,0 +1,3150 @@
+### Exceptions
+
+Any N-API function may result in a pending JavaScript exception. This is
+obviously the case for any function that may cause the execution of JavaScript
```
I think there should be a comma at the end here

mhdawson replied:
added

addaleax commented on doc/api/n-api.md:
````diff
@@ -0,0 +1,3150 @@
+and then check the value of `result`:
+
+```C
+napi_is_exception_pending()
````
Is this supposed to contain the full signature? If not, it should probably be inline code (and a link to the corresponding section)

addaleax commented on doc/api/n-api.md:
```diff
@@ -0,0 +1,3150 @@
+      break;
+    }
+    // do something with element
+  }
```
this doesn’t need to be indented

mhdawson replied:
fixed

addaleax commented on doc/api/n-api.md:
````diff
@@ -0,0 +1,3150 @@
+```
+
+##### Parameters
+- `[in] env `: The environment that the API is invoked under
````
repeated typo: space after `env`

addaleax commented on doc/api/n-api.md:
```diff
@@ -0,0 +1,3150 @@
+for a reference is 0, all subsequent calls to
+get the object associated with the reference (`napi_get_reference_value()`)
+will return NULL for the returned `napi_value`. An attempt to call
+`napi_reference_ref()` for a reference whose object has been collected,
```
The comma at the end of this line should be removed

mhdawson replied:
removed

addaleax commented on doc/api/n-api.md:
````diff
@@ -0,0 +1,3150 @@
+```
+
+For example, to define a class so that new instances can be created
+(often used with [object wrap]()):
````
object wrap? This should probably be stylized, or written as a full noun like `object wrapping`?

mhdawson replied:
change to be a link to the section describing Object Wrap

addaleax commented on doc/api/n-api.md:
```diff
@@ -0,0 +1,3150 @@
+  if (status != napi_ok) return;
+
+  status = napi_set_named_property(env, exports, "MyObject", cons);
+  if (status != napi_ok) return;
```
(also doesn’t need to be indented, )

mhdawson replied:
fixed

addaleax commented on doc/api/n-api.md:
```diff
@@ -0,0 +1,3150 @@
+N-API exposes a set of APIs to create all types of JavaScript values. 
+Some of these types are documented under 
+[Section 6](https://tc39.github.io/ecma262/#sec-ecmascript-data-types-and-values) 
+of the [ECMAScript Language Specification](https://tc39.github.io/ecma262/).
```
Can you turn this into a shorthand link and define the reference at the bottom of the file?

mhdawson replied:
Done

addaleax commented on doc/api/n-api.md:
```diff
@@ -0,0 +1,3150 @@
+1. Create a new JavaScript object
+2. Convert from a primitive C++ type to an N-API value
+3. Convert from N-API value to a primitive C++ type
+4. Get global instances including Undefined and Null
```
undefined and null should probably be lowercased and stylized as inline code pieces

mhdawson replied:
done

addaleax commented on doc/api/n-api.md:
```diff
@@ -0,0 +1,3150 @@
+##### Description
+This API allocates a `node::Buffer` object and initializes it with data copied
+from the passed-in buffer. While this is still a fully-supported data 
+structure, in most cases using a TypedArray will suffice.
```
It should probably be mentioned that in Node.js >= 4 (i.e. all currently supported versions) `Buffer`s *are* Uint8Arrays

mhdawson replied:
done.

addaleax commented on doc/api/n-api.md:
```diff
@@ -0,0 +1,3150 @@
+- `napi_ok` if the API succeeded.
+
+##### Description
+This API creates a JavaScript Symbol object from a UTF8-encoded C++ string
```
s/C++/C/

mhdawson replied:
fixed

addaleax commented on doc/api/n-api.md:
```diff
@@ -0,0 +1,3150 @@
+where each element has the same underlying binary scalar datatype.
+
+It's required that (length * size_of_element) + byte_offset should 
+be <= ArrayBuffer.ArrayBufferByteLength. If not, a RangeError exception is 
```
What is ArrayBuffer.ArrayBufferByteLength? Maybe this should read sth like `arraybuffer.byteLength`, otherwise it looks like a static property of the ArrayBuffer constructor itself

mhdawson replied:
changed to:

It's required that (length * size_of_element) + byte_offset should
be <= the size in bytes of the array passed in. If not, a RangeError exception is
raised.

addaleax commented on doc/api/n-api.md:
```diff
@@ -0,0 +1,3150 @@
+- `napi_ok` if the API succeeded.
+
+##### Description
+This API is used to convert from the C++ double type to the JavaScript 
```
s/C++/C/

mhdawson replied:
done

addaleax commented on doc/api/n-api.md:
```diff
@@ -0,0 +1,3150 @@
+ArrayBuffer. It's also safe to use the returned data buffer within the same
+callback as long as you don't call any other APIs that might trigger a GC.
+
+<TODO: What happens to this buffer in case of a compacting GC?>
```
If you’re talking about the data pointer possibly being moved: That does [not](https://github.com/nodejs/node/blob/bde26a62bfeb70b642b04bed401c03ba6f1de936/src/node.cc#L4268-L4272) happen. :)

mhdawson replied:
removed TODO.

addaleax commented on doc/api/n-api.md:
```diff
@@ -0,0 +1,3150 @@
+- `napi_boolean_expected` if a non-boolean `napi_value` is passed in
+
+##### Description
+This API retuns C++ boolean primitive equivalent of the given JavaScript
```
typo: returns, also s/C++/C/ again

mhdawson replied:
fixed, and also went through and looked for other cases as well.

addaleax commented on doc/api/n-api.md:
```diff
@@ -0,0 +1,3150 @@
+- `napi_number_expected` if a non-number `napi_value` is passed in
+
+##### Description
+This API returns the C++ double primitive equivalent of the given JavaScript
```
s/C++/C/ (going to stop commenting that now :smile:)

addaleax commented on doc/api/n-api.md:
```diff
@@ -0,0 +1,3150 @@
+- `napi_invalid_arg` if a non-external `napi_value` is passed in
+
+##### Description
+This API returns the Pointer to the data wrapped by the JavaScript
```
typo: lowercase pointer

mhdawson replied:
fixed

addaleax commented on doc/api/n-api.md:
```diff
@@ -0,0 +1,3150 @@
+
+##### Description
+This API returns the C++ int32 primitive equivalent
+of the given JavaScript Number
```
These should probably mention what happens when the value does not fit into the corresponding range?

jasnell commented on doc/api/addons.md:
```diff
@@ -213,6 +213,24 @@ Addon developers are recommended to use to keep compatibility between past and
 
+
+## N-API
+
```
Should include the `Experimental` stability index flag here.

mhdawson replied:
ok, we have that in the actual section on n-api but I can add here as well

jasnell replied:
Having it in both places is good.

jasnell commented on doc/api/n-api.md:
```diff
@@ -0,0 +1,3150 @@
+`napi_pending_exception` will be returned.
+
+In cases where a return value other than `napi_ok` or
+`napi_pending_exception` is returned, you must call
```
Please avoid the use of `you` in the docs... here and throughout.

mhdawson replied:
Did a pass removing all of the 'you's.

Fishrock123 commented on doc/api/n-api.md:
````diff
@@ -0,0 +1,3150 @@
+#### napi_throw
+##### Signature
+```C
+NODE_EXTERN napi_status napi_throw(napi_env e, napi_value error);
````
Here and throughout, can we make sure that the signature has the same names as the listed parameters?

e.g. `e` should be `env`  to match.

mhdawson replied:
fixed all of the env instances.

mhdawson commented:
@addaleax @jasnell @Fishrock123 pushed commits to address all of the comments so far.  Note that the short links render ok when you do make doc and view as html as opposed to what you see when you look at the rendered markdown link you posted.

vsemozhetbyt commented:
This is not path new doc linting:
```
  1987:1   error  Unexpected var, use let or const instead                                       no-var
  2011:1   error  Unexpected var, use let or const instead                                       no-var
  2035:1   error  Unexpected var, use let or const instead                                       no-var
  2036:1   error  Unexpected var, use let or const instead                                       no-var
  2056:1   error  Unexpected var, use let or const instead                                       no-var
  2455:1   error  Opening curly brace does not appear on the same line as controlling statement  brace-style
  2456:5   error  Expected indentation of 2 spaces but found 4                                   indent
  2546:1   error  Unexpected var, use let or const instead                                       no-var
  2630:1   error  Opening curly brace does not appear on the same line as controlling statement  brace-style
  2631:5   error  Expected indentation of 2 spaces but found 4                                   indent
  2634:1   error  Unexpected var, use let or const instead                                       no-var
  2634:11  error  Strings must use singlequote                                                   quotes
  2635:1   error  Unexpected var, use let or const instead                                       no-var

✖ 13 problems (13 errors, 0 warnings)
```
Lines like this also may need to be addressed for consistency:
```js
// var obj = {}
```

Testing Linter CI: https://ci.nodejs.org/job/node-test-linter/8536/

vsemozhetbyt commented:
Strange. Why Linting passes on CI?

vsemozhetbyt commented:
Testing if full CI also pass with linting: https://ci.nodejs.org/job/node-test-pull-request/7659/

Fishrock123 commented on doc/api/n-api.md:
````diff
@@ -0,0 +1,3024 @@
+Integral status code indicating the success or failure of a N-API call.
+Currently, the following status codes are supported.
+```C
+  napi_ok,
````
Should this be an `enum` def?

mhdawson replied:
change to be enum definition

Fishrock123 commented on doc/api/n-api.md:
```diff
@@ -0,0 +1,3024 @@
+it can be obtained by calling `napi_get_last_error_info`.
+
+### *napi_extended_error_info*
+#### Definition
```
Can we apply the same hear rules to e.g. here?

mhdawson replied:
Not sure what you mean here.

mhdawson replied:
If you meant removing the headers (Definition and Members) I went ahead and did that.  Seems better to me.

Fishrock123 commented on doc/api/n-api.md:
```diff
@@ -0,0 +1,3024 @@
+  not implemented for any VM.
+- engine_error_code: VM-specific error code. This is currently
+  not implemented for any VM.
+- error_code: the N-API status code that originated with the last error
```
```
- `member`: Regular grammar rules.
```

Ditto for the parameters section of most APIs. I think this is the style we use elsewhere.

mhdawson replied:
Not sure what you mean here either.

gibfahn replied:
I think the point is:

    error_code -> `error_code`

mhdawson replied:
Ok guessing that it meant the parameter sections use use regular grammar including  capital at the start and period at the end. Went through and fixed all that up.

mhdawson replied:
Ok, fixed error_code -> `error_code` as well.  It was pretty much ok except for the instance identified.

Fishrock123 replied:
Yeah, that's what I meant. Code blocks and grammar. Sorry.

Fishrock123 commented on doc/api/n-api.md:
````diff
@@ -0,0 +1,3024 @@
+                         const napi_extended_error_info** result);
+```
+- `[in]  env`: The environment that the API is invoked under
+- `[out]  result`: The `napi_extended_error_info` structure with more
````
Can we use 1 space between `[in/out]` and `name`? Here and throughout. They don't render with more than 1 space it seems anyways?

(Unless `make doc` does, but I doubt it?)

mhdawson replied:
Done

mhdawson commented:
@vsemozhetbyt pushed changes to address new doc lint issues, also applied your changes in https://github.com/nodejs/node/pull/12640 and then ran make jslint-ci and make jslint and no errors were reported.  Anything else I need to do ?

vsemozhetbyt commented:
@mhdawson Thank you! If nobody objects in the https://github.com/nodejs/node/pull/12640 I will land it soon and we will test with CI)

mhdawson commented:
@Fishrock123 @vsemozhetbyt believe I have addressed all of your additional comments.

mhdawson commented:
@Fishrock123 as a check of the md syntax I opened the n-api.md file in the atom editor and scanned the rendered view.  It seemed fine.

vsemozhetbyt commented:
Linter CI: https://ci.nodejs.org/job/node-test-linter/8559/

vsemozhetbyt commented:
If this line is JavaScript-ish, it also could be unified:

https://github.com/mhdawson/io.js/blame/5bb5431c049829c37c7431f02694b735070a9db2/doc/api/n-api.md#L2469

mhdawson commented:
@vsemozhetbyt pushed commit to fix that.

Fishrock123 commented:
Do the in-doc cross-reference links work with `make doc`? They appear to not parse correctly in the GitHub markdown view. (Maybe something to do with the link table at the bottom?)

mhdawson commented:
@Fishrock123 The links work fine with make doc.  Opening the md in atom, they render ok but the links don't work.  However, doing the same thing in git with other docs   (ex dns.md) the links seem to render ok but the links don't work either.  I'm guessing the rendering issue we are seeing is due to the large doc size.  The links not working is likely because we have set it up to work with make doc as opposed to md through the tables at the bottom.

jasnell commented on doc/api/n-api.md:
````diff
@@ -0,0 +1,3021 @@
+-->
+```C
+napi_status napi_instanceof(napi_env env,
+                            nap
````
Still a few more `You`s in here.

jasnell replied:
Perhaps move this link to a reference at the end of the doc and use the `[Section 9.1.6][]` syntax to avoid the long line in the middle of the paragraph

mhdawson replied:
ok must have only searched on lower case you last time.  Think I have them this time.

mhdawson replied:
done

jasnell commented on doc/api/n-api.md:
```diff
@@ -0,0 +1,3021 @@
+* [Aynchronous Operations][]
+
+The N-API is a C API that ensures ABI stability
+across Node.js versions and different compiler levels. However, we also
```
Extremely minor nit: Weird line wrapping here. Can this perhaps be reflowed a bit to even out the line lengths?

mhdawson replied:
Done

jasnell commented on doc/api/n-api.md:
```diff
@@ -0,0 +1,3021 @@
+
+
+#### napi_async_execute_callback
+
```
Nit: Inconsistent whitespace between the header and the text. Items above do not have a blank, items below do.

mhdawson replied:
ok believe I removed the extra spaces throughout.

jasnell commented on doc/api/n-api.md:
```diff
@@ -0,0 +1,3021 @@
+> Stability: 1 - Experimental
+
+
+N-API is an API for building native Addons. It is independent from
```
Suggestion: it might be beneficial to have a quick preferred pronunciation guide for `N-API` included if we really don't want people calling it "nappy"...`

mhdawson replied:
Added.

mhdawson commented:
@jasnell push commit to address your latest comments.

mhdawson commented:
@Fishrock123 @addaleax I think I've addressed all of the comments.

mhdawson commented:
@Fishrock123 github/large file may not be helping us here as I'm not 100% sure what's left with respect to https://github.com/nodejs/node/pull/12549/files#r112795649.  I already added the stability index as James requested and I think that's the comment that that one shows me.

mhdawson commented:
@Fishrock123 I don't have any great ideas on the headings right now, so I'd prefer to land this and then I can spend a bit more time and maybe come up with something in a later PR.

mhdawson commented:
@addaleax I think I've addressed your issues so I plan to land soon.  If not just let me know.

CI run:  https://ci.nodejs.org/job/node-test-pull-request/7734/

addaleax commented:
@mhdawson Yeah, sorry, didn’t have the time to give it a full review. So please don’t let me hold up anything! :)

mhdawson commented:
@addaleax thanks. Of course if you spot anything later on just let me know and I'll cover it in a later PR.

mhdawson commented:
Ok CI good, landing.

mhdawson commented:
landed as deb9622b11a2780f35e5d122273abcf6458bbe4b

refack commented:
FYI: improve on #12898 when this lands

refack commented:
NM
